Add a health configuration option of which alarms to load

MrZammler opened the pull request:
##### Summary
<!--
Describe the change in summary section, including rationale and design decisions.
Include "Fixes #nnn" if you are fixing an existing issue.
-->

This PR adds a health configuration option called `enabled alarms` which, using simple patterns, controls which alerts to load.

By default (`*`) all alerts are loaded.

It can be used to disable specific alerts if the user wants it. It provides another way to disable alerts apart from silencing their notifications (silenced alerts will still show up in e.g. cloud).

##### Test Plan

<!--
Provide enough detail so that your reviewer can understand which test cases you
have covered, and recreate them if necessary. If our CI covers sufficient tests, then state which tests cover the change.
-->

Ensure without changing the configuration, that all alerts are loaded normally.

Test the configuration option with various cases:

1) `*` -> Should load all alerts.
2) `!*oom* *` -> Should load all alerts except `oom_killer`.
3) `oom_killer` -> Should only load `oom_killer`.

##### Additional Information
<!-- This is usually used to help others understand your
motivation behind this change. A step-by-step reproduction of the problem is
helpful if there is no related issue. -->

This will also have an effect on the alerts loaded for children. We could do a follow up PR which could add this on a per-host level in `stream.conf`.

<details> <summary>For users: How does this change affect me?</summary>
  <!--
Describe the PR affects users: 
- Which area of Netdata is affected by the change?
- Can they see the change or is it an under the hood? If they can see it, where?
- How is the user impacted by the change? 
- What are there any benefits of the change? 
-->
</details>

andrewm4894 commented:
Is it "alarms" or "alerts"? I think i recall @amalkov saying we were going to settle on "Alerts" but unsure if thats in context of NC or not. 

Maybe its a different thing to standardize another time perhaps or if its "Alarms" or "Alerts" or if there is a difference and reason we may need to keep both terms.

MrZammler commented:
Hi @andrewm4894 yeah, initially I called it `enabled alerts`, however currently throughout the agent's documentation we call them `alarms` (mostly) and I didn't want to differ there.

We need to maybe do a big effort to change the name all around.

ilyam8 commented on daemon/config/README.md:
```diff
@@ -175,6 +175,7 @@ monitoring](/health/README.md).
 |           run at least every seconds           |                       `10`                       | Controls how often all alarm conditions should be evaluated.                                                                                                                                   |
 | postpone alarms during hibernation for seconds |                       `60`                       | Prevents false alarms. May need to be increased if you get alarms during hibernation.                                                                                                          |
 |             rotate log every lines             |                       2000                       | Controls the number of alarm log entries stored in `<lib directory>/health-log.db`, where `<lib directory>` is the one configured in the [\[global\] section](#global-section-options)         |
+|                enabled alarms                  |                       *                          | Defines which alarms to load, both from stock and custom directories. This is a [simple pattern](/libnetdata/simple_pattern/README.md) list of alarm or template names. Can be used to disable specific alarms, for example `enabled alarms =  !oom_kill *` will load all alarms except `oom_kill`. |
```
```suggestion
|                enabled alarms                  |                       *                          | Defines which alarms to load from both standard and custom directories. This is a [simple pattern](/libnetdata/simple_pattern/README.md) list of alarm or template names. Can be used to disable specific alarms. For example, `enabled alarms =  !oom_kill *` will load all alarms except `oom_kill`. |
```

ilyam8 replied:
yeah, changed it to _user and custom_.

ilyam8 replied:
And it is a suggestion. But I'm sure we need to simplify the last sentence (with "for example") by dividing it into two.

andrewm4894 commented:
@MrZammler should we add this to 
PH telemetry events? thinking being it could be useful for surfacing alerts that are commonly disabled and so maybe need tweaking.

MrZammler commented:
> @MrZammler should we add this to PH telemetry events? thinking being it could be useful for surfacing alerts that are commonly disabled and so maybe need tweaking.

Sure, yes! I will add it as part of this PR.

MrZammler commented:
Hi @andrewm4894 

I've added it to the analytics as `health_conf_alarms_enabled`, it will be exactly as in the config (i.e. as a simple pattern). Would we be able to make sense of it? (It could be e.g. `!oom_killer *`).

Also added `health_enabled` -> true/false if global health is enabled or not.

andrewm4894 commented:
> Would we be able to make sense of it? (It could be e.g. `!oom_killer *`).

Cool, yep will be fine like that and can wrangle further downstream

MrZammler commented:
@andrewm4894 well, I've changed it to a list of disabled alarms. I think it would make life much easier.

Also added a list of silenced alarms. So now we should have e.g.:

```
"health_alarms_silenced": "|system_clock_sync_state|1min_netdev_budget_ran_outs|lowest_entropy|10min_cpu_usage|10min_disk_backlog|10min_disk_utilization"
"health_conf_alarms_disabled": "|ram_in_use|ram_available|30min_ram_swapped_out|cgroup_ram_in_use|k8s_cgroup_ram_in_use|wmi_ram_in_use|"

```

I think it's better like this?

ilyam8 commented:
> thinking being it could be useful for surfacing alerts that are commonly disabled and so maybe need tweaking.

I personally doubt this source can/will be used to make a decision about tweaking. 

@MrZammler how about opening a separate PR with PH events so it is not a blocker for the functionality to get merged?

ilyam8 commented:
@MrZammler found a bug:

- I have `enabled alarms = !10min_disk_backlog *`
- And on every restart I see the following in the logs:

```bash
2023-06-05 22:32:59: netdata ERROR : HEALTH : Health configuration at line 155 of file '/opt/netdata/usr/lib/netdata/conf.d/health.d/disks.conf' has unknown key 'on'. Expected either 'alarm' or 'template'. (errno 2, No such file or directory)
2023-06-05 22:32:59: netdata ERROR : HEALTH : Health configuration at line 156 of file '/opt/netdata/usr/lib/netdata/conf.d/health.d/disks.conf' has unknown key 'class'. Expected either 'alarm' or 'template'.
2023-06-05 22:32:59: netdata ERROR : HEALTH : Health configuration at line 157 of file '/opt/netdata/usr/lib/netdata/conf.d/health.d/disks.conf' has unknown key 'type'. Expected either 'alarm' or 'template'.
2023-06-05 22:32:59: netdata ERROR : HEALTH : Health configuration at line 158 of file '/opt/netdata/usr/lib/netdata/conf.d/health.d/disks.conf' has unknown key 'component'. Expected either 'alarm' or 'template'.
2023-06-05 22:32:59: netdata ERROR : HEALTH : Health configuration at line 161 of file '/opt/netdata/usr/lib/netdata/conf.d/health.d/disks.conf' has unknown key 'lookup'. Expected either 'alarm' or 'template'.
2023-06-05 22:32:59: netdata ERROR : HEALTH : Health configuration at line 162 of file '/opt/netdata/usr/lib/netdata/conf.d/health.d/disks.conf' has unknown key 'units'. Expected either 'alarm' or 'template'.
2023-06-05 22:32:59: netdata ERROR : HEALTH : Health configuration at line 163 of file '/opt/netdata/usr/lib/netdata/conf.d/health.d/disks.conf' has unknown key 'every'. Expected either 'alarm' or 'template'.
2023-06-05 22:32:59: netdata ERROR : HEALTH : Health configuration at line 164 of file '/opt/netdata/usr/lib/netdata/conf.d/health.d/disks.conf' has unknown key 'warn'. Expected either 'alarm' or 'template'.
2023-06-05 22:32:59: netdata ERROR : HEALTH : Health configuration at line 165 of file '/opt/netdata/usr/lib/netdata/conf.d/health.d/disks.conf' has unknown key 'delay'. Expected either 'alarm' or 'template'.
2023-06-05 22:32:59: netdata ERROR : HEALTH : Health configuration at line 166 of file '/opt/netdata/usr/lib/netdata/conf.d/health.d/disks.conf' has unknown key 'info'. Expected either 'alarm' or 'template'.
2023-06-05 22:32:59: netdata ERROR : HEALTH : Health configuration at line 167 of file '/opt/netdata/usr/lib/netdata/conf.d/health.d/disks.conf' has unknown key 'to'. Expected either 'alarm' or 'template'
```

I think we need to silently ignore the rest of the alarm configuration if it has been filtered out.
